### ISSUE TITLE
Unable to save modified files (TypeError: t._dataview is undefined)

### ISSUE THREAD

**Issue by MHLoppy** [gmariani/minerva#88]
When saving a .sol file after making any changes, I receive the following error:

```
TypeError: t._dataview is undefined
(https://mariani.life/projects/minerva/js/lib/ByteArray.20200510_1589162844974.js:1)
```

The only time I don't receive this error is if I open a file and then save it before changing anything (in which case it appears to give me the same file I gave it to start with). Exporting still works fine, and the exported data file appears to work as expected.

Attached is a sample file that I've tried to edit then save, as I'm not sure if the file itself matters (I only have one game's .sol files to test with). I encountered this problem on both Firefox 76.0 and Brave 1.9.72.

[rebuild3_fort_11.zip](https://github.com/gmariani/minerva/files/4678470/rebuild3_fort_11.zip)

**Comment by 3plus4i** [comment]
I encountered the exact same issue with Rebuild 2. Since there was no answer for almost two years, I guess editing the swf to export the data in an editable way it is then.

**Comment by gmariani** [comment]
TBH I forgot about this bug, and haven't worked on the project in a while (over two years apparently). I can try taking a look at it again and see whats going on.

**Comment by gmariani** [comment]
I rebuilt the app, and that alone appears to have fixed the issue if you want to give it a try.

**Comment by 3plus4i** [comment]
Unfortunately the issue persists for me. Saves might not be the same for Rebuild 2 and 3, so I'll attach one. There's also two things that seem odd to me:

1. according to the game code, the save should simply be zlib encoded xml as a bytearray, but after decoding, the first 4 bytes don't look like they belong there ![scr151](https://user-images.githubusercontent.com/12765347/158876002-1159b682-7012-4bac-a93c-8ad4b4329211.png)
2. when I export as JSON, the data looks fine, but after reimporting the edited file, it looks like hot garbage ![scr152](https://user-images.githubusercontent.com/12765347/158876170-cda60e2d-5404-46ac-8055-f8aa26604c93.png)

[Rebuild2_Fort_0.zip](https://github.com/gmariani/minerva/files/8289415/Rebuild2_Fort_0.zip)

**Comment by gmariani** [comment]
"according to the game code" - do you have the source code available? If so, please send it over, I'd like to see what they are trying to do here.

My best guess is that it's double encoded as a bytearray or some other datatype and those four bytes describe the payload. 

But you stumbled across another bug/typo that has been there since 2014... Apparently not many people use the import bytearray feature. When you imported, it wasn't displaying the file you imported but the main file itself.... sigh.

Well now you can export and import 1:1, so I just need to solve the mystery of the four bytes. I will say that in my sample you gave me the four bytes are different from the screenshot you sent which makes me more confident that it's describing possibly the type and length of the data. Let me know if you do have any of the source code by chance, thanks.

**Comment by 3plus4i** [comment]
Those four bytes are different on every save. I don't have the source code, I decompiled the game with JPEXS. I'll append the two .as files that *should* be relevant. The functions should be `saveFort()` in both files, with the one in `SaveManager.as` building the XML. There is also obfuscation going on, but it should mostly only consist of dead code.

[rebuild_2_source.zip](https://github.com/gmariani/minerva/files/8297868/rebuild_2_source.zip)

**Comment by gmariani** [comment]
I got it figured out. I believe they are encoding the raw XML data into a byte array, and then saving that into the shared object which again encodes it as a bytearray. The extra four bytes you're seeing is this: the first byte is the XML marker (11), the next 3 bytes just happen to be similar, but they are actually a UInt29 which can be 1-4 bytes in length. That UInt29 is the length of the XML string to follow, so it follows the AMF3 spec exactly, which is why I suspect it was double encoded for whatever reason. But this makes it a bit difficult to edit because any change to the length means the header needs to be updated.

```
Rebuild2_Fort_0.sol
0x0b 0x83 0xf6 0x05
Length 32130
0x0b = 11 <-- XML Marker
0x83 = 131 <-- UInt29
(131 & 0x7f) << 7 = 384
0xf6 = 246 <-- UInt29
(384 | (246 & 0x7f)) << 7 = 64256
0x05 = 5 <-- UInt29
64256 | 5 = 64261
64261 & 1 = 1
64261 >> 1 = 32130 <-- Matches length of XML
```

```
rebuild3_fort_11.sol
0x0b 0x9e 0xce 0x05
Length 250754
0x0b = 11 <-- XML Marker
0x9e = 158 <-- UInt29
(158 & 0x7f) << 7 = 3840
0xce = 206 <-- UInt29
(3840 | (206 & 0x7f)) << 7 = 501504
0x05 = 5 <-- UInt29
501504 | 5 = 501509
501509 & 1 = 1
501509 >> 1 = 250754 <-- Matches length of XML
```
Now, if you're willing to work a little bit, you can still get around this. Just extract the bytearray, and remove the header. Then create a new sharedobject with just the XML. Edit that XML in minerva and save it. You'll then be able to see what the correct header should be when looking at the bytes it generated. Here is a sample file where you can see the bytes preceding the XML matches
[test_xml.zip](https://github.com/gmariani/minerva/files/8301376/test_xml.zip)
.

**Comment by gmariani** [comment]
Ah, reviewing their code I see what their logic was. Since the XML files can be lengthy, they were zipping them up before storing them in the shared object as a way to keep save file sizes down.

**Comment by 3plus4i** [comment]
Thanks for your work! I did not realize that it was converted to a bytearray twice.

So if I understand this right, it's only the length value that needs to adjusted, if the length changed. If I know the length I could edit the bytes manually, can't I?

**Comment by gmariani** [comment]
Yes, only the length value would need to be adjusted after you edit the XML content itself. UInt29 was a 'clever' invention by Adobe to write an integer using only the maximum number of bytes as necessary. So if it's a small number, it will use one byte, etc up until a maximum of 4 bytes. It's a complicated way to calculate the integer value, so the best bet would be to let minerva do the heavy lifting.

I'd suggest using that test_xml.sol i attached above to edit the XML. You will then open the .sol file in a hex editor where you can see the bytes preceding your XML. I only suggest this because I know you know how to do it already based on your screenshots. So you would look for the start of the XML and go left until you see the 0x0B byte (the XML marker). Then copy from that marker to the end of the XML and paste that as your byte array data. It's a bit of a hassle but it will get the job done. Attached a screenshot too.

![Capture](https://user-images.githubusercontent.com/5058619/159107427-b47e68f3-c823-471f-8799-60cff010c97c.JPG)

For reference, here is how to read/write a UInt29:

https://github.com/gmariani/minerva/blob/af0adac9c331448fc3fa48f8db5f910a40a2f3a2/app/js/lib/AMF3.js#L343-L410

**Comment by 3plus4i** [comment]
It was about time that I actually tested it again. Unfortunately I stumbled upon a few issues.

The dialog to import JASON into the test_xml.sol seems a bit odd. It has a filed for a file name but doesn't open an open file dialogue. A file name is required to import JASON though and if I put something in it says that the JASON is invalid, although I just copied the text from the file I exported to JASON and edited. Maybe I'm misunderstanding the import. Anyway, putting the text directly into the XML item worked.

I couldn't figure out how to copy the byte representation from my HEX editor, but it worked by importing the test_xml.sol as bytearray and removing everything up until the XML marker. Saving that works, but I have to compress it with zlib first. After doing that, I get the same error about t._dataview again. There also is a `00` byte at the end after importing, after the closing bracket of </data>, which wasn't there in the original data. It makes no difference though if I keep it in or remove it, at least so far.

**Comment by 3plus4i** [comment]
I tried to debug it, but I keep hitting huge roadblocks. I found the code for the allegedly missing Dataview, though it makes no sense to me how it could be missing, since it's created right there. I found no way of accessing the serviceworker context at runtime to run code step by step, so I seem to be at a dead end again.
